In [1]:
import numpy as np
import pandas as pd
from sklearn.linear_model import LinearRegression

__Задача 1__

Дана матрица объект-признак
```
X = [[ 1.22401313,  2.30868478,  3.03636353,  2.69287214],
     [-0.18757272,  1.30337355,  5.12093014,  3.46363202],
     [-0.81094525,  1.82463398,  5.79686488,  1.86159445],
     [ 0.75129018,  2.67392052,  3.65529809,  1.66746094],
     [ 0.00972362,  1.97367255,  2.50594319,  1.69755173],
     [-0.62972637,  0.77750764,  2.84124027,  4.54410559],
     [ 2.29536229,  1.81206697,  1.95026215,  1.51874636],
     [ 0.0920418 ,  2.26971361,  7.47708735,  2.61081203],
     [ 2.39252799,  3.17563985,  3.61420599,  5.10773362],
     [ 0.54983815,  2.87988651,  1.65752765,  1.59635987]]
```
и значения целевой переменной
```
y = [ 9.26193358,  9.700363  ,  8.67214805,  8.74796974,  6.18689108,
      7.53312713,  7.57643777, 12.44965478, 14.29010746,  6.68361218]
```

Подберите два признака (из четырёх) так, чтобы уровень линейной зависимости целевой переменной от значений этих признаков был максимальным. Другими словами, модель линейной регрессии на этих признаках должна давать наилучший результат. В качестве ответа запишите значение коэффициента детерминации для модели на этих признаках.


In [2]:
X = np.array([[ 1.22401313,  2.30868478,  3.03636353,  2.69287214],
     [-0.18757272,  1.30337355,  5.12093014,  3.46363202],
     [-0.81094525,  1.82463398,  5.79686488,  1.86159445],
     [ 0.75129018,  2.67392052,  3.65529809,  1.66746094],
     [ 0.00972362,  1.97367255,  2.50594319,  1.69755173],
     [-0.62972637,  0.77750764,  2.84124027,  4.54410559],
     [ 2.29536229,  1.81206697,  1.95026215,  1.51874636],
     [ 0.0920418 ,  2.26971361,  7.47708735,  2.61081203],
     [ 2.39252799,  3.17563985,  3.61420599,  5.10773362],
     [ 0.54983815,  2.87988651,  1.65752765,  1.59635987]])

y = np.array([ 9.26193358,  9.700363  ,  8.67214805,  8.74796974,  6.18689108,
      7.53312713,  7.57643777, 12.44965478, 14.29010746,  6.68361218])

X = X[:, :3]

X.shape, y.shape

((10, 3), (10,))

In [3]:
ones = np.ones((X.shape[0], 1))
X = np.hstack((ones, X))
X

array([[ 1.        ,  1.22401313,  2.30868478,  3.03636353],
       [ 1.        , -0.18757272,  1.30337355,  5.12093014],
       [ 1.        , -0.81094525,  1.82463398,  5.79686488],
       [ 1.        ,  0.75129018,  2.67392052,  3.65529809],
       [ 1.        ,  0.00972362,  1.97367255,  2.50594319],
       [ 1.        , -0.62972637,  0.77750764,  2.84124027],
       [ 1.        ,  2.29536229,  1.81206697,  1.95026215],
       [ 1.        ,  0.0920418 ,  2.26971361,  7.47708735],
       [ 1.        ,  2.39252799,  3.17563985,  3.61420599],
       [ 1.        ,  0.54983815,  2.87988651,  1.65752765]])

In [4]:
XTX = X.T.dot(X)

In [5]:
XTX_inv = np.linalg.inv(XTX)

b = XTX_inv.dot(X.T).dot(y)
b

array([3.05093578, 1.47184411, 0.35994664, 1.18613078])

In [6]:
z = X.dot(b)
e = y - z

e

array([-0.02308628,  0.38226725, -0.71781345, -0.70687833, -0.56114949,
        1.75910661, -1.81842685, -0.42253139,  2.2877616 , -0.17924967])

In [7]:
e.mean()

2.957634137601417e-14

In [25]:
# через skliearn
lr = LinearRegression(fit_intercept=False).fit(X, y)
b, lr.coef_

(array([3.05093578, 1.47184411, 0.35994664, 1.18613078]),
 array([3.05093578, 1.47184411, 0.35994664, 1.18613078]))

In [10]:
def sum_of_squares(samples: np.ndarray) -> float:
    """Сумма квадратов отклонений.
    """
    
    return ((samples - samples.mean()) ** 2).sum()

In [11]:
R = 1 - sum_of_squares(e) / sum_of_squares(y)
R

# 0.7701157002858928    максимальный R при применении первого и второго признака

0.7701157002858928

In [12]:
np.corrcoef(y, z) ** 2

array([[1.       , 0.7701157],
       [0.7701157, 1.       ]])

__Задание 2__

Является ли значимым уравнение регрессии, полученное в предыдущей задаче? В качестве ответа запишите 1, если является, и 0 иначе.

In [13]:
n = X.shape[0]
m = 2

k1 = m
k2 = n - m - 1

k1, k2

(2, 7)

In [14]:
from scipy import stats

In [15]:
alpha = 0.05

t = stats.f.ppf(1 - alpha, k1, k2)
t

4.73741412777588

In [16]:
F = (R / k1) / ((1 - R) / k2)
F

11.725050185474748

11.72505 > 4.7374  --->  уравнение регрессии значимо

Ответ 1

__Задача 3__

Для проведения A/B-тестирования сайта интернет-магазина были получены следующие данные: страница `A` была посещена 2509 раз, из них 77 закончились совершением покупки, страница `B` была посещена 1465 раз, 60 из них закончились совершением покупки. Является ли значимым отличие конверсии на страницах `A` и `B`? В качестве ответа запишите 1, если является, и 0 иначе.

_Подсказка_. Реализуйте двухвыборочный t-тест. В качестве выборок здесь можно взять наборы меток совершения покупки (0 или 1) каждым посетителем.

In [17]:
A = np.array([1 for _ in range(77)] + [0 for _ in range(2509-77)])
A

array([1, 1, 1, ..., 0, 0, 0])

In [18]:
B = np.array([1 for _ in range(60)] + [0 for _ in range(1465-60)])
B

array([1, 1, 1, ..., 0, 0, 0])

In [19]:
n1 = A.size
n2 = B.size

s1 = A.std(ddof=1)
s2 = B.std(ddof=1)

In [20]:
s_delta = np.sqrt(s1 ** 2 / n1 + s2 ** 2 / n2)
s_delta

0.006220171278295827

In [21]:
t = (A.mean() - B.mean()) / s_delta
t

-1.6504551408398205

In [22]:
df = (s1 ** 2 / n1 + s2 ** 2 / n2) ** 2 / \
    ((s1 ** 2 / n1) ** 2 / (n1 - 1) + (s2 ** 2 / n2) ** 2 / (n2 - 1))

df

2732.8025644352133

In [23]:
alpha = 0.05

t1 = stats.t.ppf(alpha / 2, df=df)
t2 = stats.t.ppf(1 - alpha / 2, df=df)

t1, t2

(-1.9608324352746576, 1.9608324352746571)

значение t попало в критическую область --->> конверсия в тестах не отличается

In [24]:
# провека через stat
stats.ttest_ind(A, B, equal_var=False)

Ttest_indResult(statistic=-1.6504551408398205, pvalue=0.09896479217925898)

Ответ: 0